x-pack/metricbeat/module/aws/billing/billing.go

```diff
@@ -0,0 +1,387 @@
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	logger := logp.NewLogger(metricsetName)
+	metricSet, err := aws.NewMetricSet(base)
+	if err != nil {
+		return nil, fmt.Errorf("error creating aws metricset: %w", err)
+	}
+
+	config := struct {
+		CostExplorerConfig CostExplorerConfig `config:"cost_explorer_config"`
+	}{}
+
+	err = base.Module().UnpackConfig(&config)
+	if err != nil {
+		return nil, fmt.Errorf("error unpack raw module config using UnpackConfig: %w", err)
+	}
+
+	logger.Debugf("cost explorer config = %s", config)
+
+	return &MetricSet{
+		MetricSet:          metricSet,
+		logger:             logger,
+		CostExplorerConfig: config.CostExplorerConfig,
+	}, nil
+}
+
+// Validate checks if given dimension keys are supported.
+func (c CostExplorerConfig) Validate() error {
+	for _, key := range c.GroupByDimensionKeys {
+		supported, _ := aws.StringInSlice(key, supportedDimensionKeys)
+		if !supported {
+			return fmt.Errorf("costexplorer GetCostAndUsageRequest does not support dimension key: %s", key)
+		}
+	}
+	return nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) error {
+	// Get startDate and endDate
+	startDate, endDate := getStartDateEndDate(m.Period)
+
+	// Get startTime and endTime
+	startTime, endTime := aws.GetStartTimeEndTime(m.Period)
+
+	// get cost metrics from cost explorer
+	awsConfig := m.MetricSet.AwsConfig.Copy()
+	svcCostExplorer := costexplorer.New(awscommon.EnrichAWSConfigWithEndpoint(
+		m.Endpoint, "monitoring", "", awsConfig))
+
+	awsConfig.Region = regionName
+	svcCloudwatch := cloudwatch.New(awscommon.EnrichAWSConfigWithEndpoint(
+		m.Endpoint, "monitoring", regionName, awsConfig))
+
+	timePeriod := costexplorer.DateInterval{
+		Start: awssdk.String(startDate),
+		End:   awssdk.String(endDate),
+	}
+
+	var events []mb.Event
+
+	// Get estimated charges from CloudWatch
+	eventsCW := m.getCloudWatchBillingMetrics(svcCloudwatch, startTime, endTime)
+	events = append(events, eventsCW...)
+
+	// Get total cost from Cost Explorer GetCostAndUsage with group by type "DIMENSION" and "TAG"
+	eventsCE := m.getCostGroupBy(svcCostExplorer, m.CostExplorerConfig.GroupByDimensionKeys, m.CostExplorerConfig.GroupByTagKeys, timePeriod, startDate, endDate)
+	events = append(events, eventsCE...)
+
+	// report events
+	for _, event := range events {
+		if reported := report.Event(event); !reported {
+			m.Logger().Debug("Fetch interrupted, failed to emit event")
+			return nil
+		}
+	}
+	return nil
+}
+
+func (m *MetricSet) getCloudWatchBillingMetrics(svcCloudwatch cloudwatchiface.ClientAPI, startTime time.Time, endTime time.Time) []mb.Event {
```

Comment:
nit: I would consider formating this header's variables line-by-line, sth like:

```suggestion
func (m *MetricSet) getCloudWatchBillingMetrics(
svcCloudwatch cloudwatchiface.ClientAPI, 
startTime time.Time,
 endTime time.Time) []mb.Event {
```

Reply:
done